lib/pilot/build_manager.rb

```diff
@@ -58,45 +58,69 @@ def describe_build(build)
+    def obtain_uploaded_build
+      # the upload date of the new build
       # we use it to identify the build
+      # or use config[:build_full_version] if provided
 
       start = Time.now
-      wait_processing_interval = config[:wait_processing_interval].to_i
-      latest_build = nil
-      loop do
-        Helper.log.info "Waiting for iTunes Connect to process the new build"
-        sleep wait_processing_interval
-        builds = app.all_processing_builds
-        break if builds.count == 0
-        latest_build = builds.last # store the latest pre-processing build here
+
+      processing_build = wait_for_processing_build
+      full_build = wait_for_full_build(processing_build)
+
+      if full_build
+        minutes = ((Time.now - start) / 60).round
+        Helper.log.info "Successfully finished processing the build".green
+        Helper.log.info "You can now tweet: "
+        Helper.log.info "iTunes Connect #iosprocessingtime #{minutes} minutes".yellow
+        return full_build
+      else
+        raise "Error: Seems like iTunes Connect didn't properly pre-process the binary".red
       end
+    end
 
+    def wait_for_full_build(processing_build)
       full_build = nil
 
       while full_build.nil? || full_build.processing
         # Now get the full builds with a reference to the application and more
-        # As the processing build from before doesn't have a refernece to the application
-        full_build = app.build_trains[latest_build.train_version].builds.find do |b|
-          b.build_version == latest_build.build_version
+        # As the processing build from before doesn't have a reference to the application
+
+        full_build = app.build_trains[processing_build.train_version].builds.find do |b|
+          b.build_version == processing_build.build_version
         end
 
         Helper.log.info "Waiting for iTunes Connect to finish processing the new build (#{full_build.train_version} - #{full_build.build_version})"
         sleep wait_processing_interval
       end
 
-      if full_build
-        minutes = ((Time.now - start) / 60).round
-        Helper.log.info "Successfully finished processing the build".green
-        Helper.log.info "You can now tweet: "
-        Helper.log.info "iTunes Connect #iosprocessingtime #{minutes} minutes".yellow
-        return full_build
+      full_build
+    end
+
+    # This method will takes care of checking for the processing builds every few seconds
+    # @return [Build] The build that we just uploaded
+    def wait_for_processing_build
+      processing_build = nil
+
+      if could_determine_uploaded_build?
+        # stop if there is not build with valid version in pre-processing builds
+        should_stop = lambda {|builds| builds.select{|build| is_uploaded_build(build)}.count == 0 }
+        # store pre-processing build with valid version here
+        obtain_processing_build = lambda {|builds| builds.find{|build| is_uploaded_build(build)} }
```

Comment:
Prefer `detect` over `find`.<br>Use the new lambda literal syntax `->(params) {...}`.<br>Space missing to the left of {.